src/t_hash.c

```diff
@@ -1053,6 +1053,59 @@ void hdelCommand(client *c) {
     addReplyLongLong(c, deleted);
 }
 
+void hgetdelCommand(client *c) {
+    int fields_index = 4;
+    int i, deleted = 0;
+    long long num_fields = 0;
+    bool keyremoved = false;
+
+    if (getLongLongFromObjectOrReply(c, c->argv[fields_index - 1], &num_fields, NULL) != C_OK) return;
+
+    /* Check that the parsed fields number matches the real provided number of fields */
+    if (!num_fields || num_fields != (c->argc - fields_index)) {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+
+    /* Don't abort when the key cannot be found. Non-existing keys are empty
+     * hashes, where HGETDEL should respond with a series of null bulks. */
+    robj *o = lookupKeyWrite(c->db, c->argv[1]);
+    if (checkType(c, o, OBJ_HASH)) return;
+
+    /* Reply with array of values */
+    addReplyArrayLen(c, num_fields);
+    for (i = fields_index; i < c->argc; i++) {
+        addHashFieldToReply(c, o, c->argv[i]->ptr);
+    }
+
+    /* If hash doesn't exist, we're done as we have already replied with NULLs */
+    if (o == NULL) return;
+
+    /* Now delete the fields. */
+    bool hash_volatile_items = hashTypeHasVolatileFields(o);
+    for (i = fields_index; i < c->argc; i++) {
+        if (hashTypeDelete(o, c->argv[i]->ptr)) {
+            deleted++;
+            if (hashTypeLength(o) == 0) {
+                if (hash_volatile_items) dbUntrackKeyWithVolatileItems(c->db, o);
+                dbDelete(c->db, c->argv[1]);
+                keyremoved = true;
+                break;
+            }
+        }
+    }
```

Comment:
I think it is fine to avoid over optimization in favor of simplicity in this case.
I mean optimal implementation would be to "pop" the element from the hashtable take the value and free the memory (in case encoding is hashtable). I do think that we can improve by combining this loop with the following one eg (didn't test it :)):
```suggestion
    /* Reply with array of values */
    addReplyArrayLen(c, num_fields);
    for (i = fields_index; i < c->argc; i++) {
        addHashFieldToReply(c, o, c->argv[i]->ptr);
         /* If hash doesn't exist or the field is not found we just reply with NULL */
        if (o == NULL || !hashTypeDelete(o, c->argv[i]->ptr)) continue;
        deleted++;
        if (hashTypeLength(o) == 0) {
            if (hash_volatile_items) dbUntrackKeyWithVolatileItems(c->db, o);
            dbDelete(c->db, c->argv[1]);
            keyremoved = true;
            o = NULL;
        }
    }
```

Reply:
Yeah this should work, I will add this and test it. Thank you so much!